pkg/vere/main.c

```diff
@@ -1726,6 +1727,173 @@ _cw_prep(c3_i argc, c3_c* argv[])
   u3_Host.ops_u.tem = c3y;
 }
 
+/* _cw_chop(): truncate event log
+*/
+static void
+_cw_chop(c3_i argc, c3_c* argv[])
+{
+  c3_i ch_i, lid_i;
+  c3_w arg_w;
+
+  static struct option lop_u[] = {
+    { "loom", required_argument, NULL, c3__loom },
+    { NULL, 0, NULL, 0 }
+  };
+
+  u3_Host.dir_c = _main_pier_run(argv[0]);
+
+  while ( -1 != (ch_i=getopt_long(argc, argv, "", lop_u, &lid_i)) ) {
+    switch ( ch_i ) {
+      case c3__loom: {
+        c3_w lom_w;
+        c3_o res_o = _main_readw(optarg, u3a_bits + 3, &lom_w);
+        if ( (c3n == res_o) || (lom_w < 20) ) {
+          fprintf(stderr, "error: --loom must be >= 20 and <= %u\r\n", u3a_bits + 2);
+          exit(1);
+        }
+        u3_Host.ops_u.lom_y = lom_w;
+      } break;
+
+      case '?': {
+        fprintf(stderr, "invalid argument\r\n");
+        exit(1);
+      } break;
+    }
+  }
+
+  //  argv[optind] is always "chop"
+  //
+
+  if ( !u3_Host.dir_c ) {
+    if ( optind + 1 < argc ) {
+      u3_Host.dir_c = argv[optind + 1];
+    }
+    else {
+      fprintf(stderr, "invalid command, pier required\r\n");
+      exit(1);
+    }
+
+    optind++;
+  }
+
+  if ( optind + 1 != argc ) {
+    fprintf(stderr, "invalid command\r\n");
+    exit(1);
+  }
+
+  // gracefully shutdown the pier if it's running
+  u3_disk* log_u = _cw_disk_init(u3_Host.dir_c);
+  u3_disk_exit(log_u);
+
+  u3m_boot(u3_Host.dir_c, (size_t)1 << u3_Host.ops_u.lom_y);
+  u3e_backup(c3y);  //  backup snapshot
+  u3m_stop();
+
+  // initialize the lmdb environment
+  // see disk.c L885
+  const size_t siz_i =
+  // 500 GiB is as large as musl on aarch64 wants to allow
+  #if (defined(U3_CPU_aarch64) && defined(U3_OS_linux))
+    0x7d00000000;
+  #else
+    0x10000000000;
+  #endif
+  c3_c log_c[8193];
+  snprintf(log_c, 8192, "%s/.urb/log", u3_Host.dir_c);
+  MDB_env* env_u = u3_lmdb_init(log_c, siz_i);
+
+  // get the first/last event numbers
+  c3_d fir_d, las_d;
+  if ( c3n == u3_lmdb_gulf(env_u, &fir_d, &las_d) ) {
+    fprintf(stderr, "disk: failed to load latest event from database\r\n");
+    u3_king_bail();
+  }
+
+  // get the metadata
+  MDB_val ver;
+  if ( c3y != u3_lmdb_read_meta_one(env_u, &ver, "version") ) {
+    fprintf(stderr, "chop: failed to read version\r\n");
+    u3_king_bail();
+  }
+  c3_d ver_d = *(c3_d*)ver.mv_data;
```

Comment:
The value stored under the `"version"` key doesn't appear to be a double word. Instead, it looks either an unsigned 8-bit or unsigned 32-bit value.

Reply:
Good catch.  @matthew-levan I thought in a previous version of this code, you just left the metadata values as buffers, without converting them.  What's the reason to convert them to some more specific datatype?  I thought we just needed to write the same values into the new db.  Or am I remembering wrong?

Even simpler: can we not just use `u3_disk_read_meta()` and `u3_disk_save_meta()` directly?  Something like this:

```C
{
  u3_disk* log_u;
  c3_d     who_d[2];
  c3_o     fak_o;
  c3_w     lif_w;
  
  if ( c3n == u3_disk_read_meta(log_u, who_d, &fake_o, &lif_w) ) {
    //  error handling
  }
  
  if ( c3n == u3_disk_save_meta(log_u, who_d, fake_o, lif_w) ) {
    //  error handling
  }
}
```

We'd need to do the read and save in the same function to avoid sticking these values in some heap-allocated struct, but I think that's doable here.